File: backend/app/models/control_operativo.py
from sqlalchemy import Column, Integer, String, Text, DateTime, ForeignKey, Boolean, Date, Numeric
from sqlalchemy.orm import relationship
from datetime import datetime, timezone, timedelta
from app.config.database import Base

def get_colombia_time():
    """Retorna la fecha/hora actual en zona horaria de Colombia (UTC-5)"""
    colombia_tz = timezone(timedelta(hours=-5))
    return datetime.now(colombia_tz).replace(tzinfo=None)

class ControlOperativo(Base):
    __tablename__ = "controles_operativos"

    # ID y metadatos
    id = Column(Integer, primary_key=True, index=True)
    
    # I. DATOS DEL USUARIO
    ciudad = Column(String(100), nullable=False, default="Bogotá D.C")
    fecha_dia = Column(Integer, nullable=False)
    fecha_mes = Column(Integer, nullable=False)
    fecha_ano = Column(Integer, nullable=False)
    nombre_docente_responsable = Column(String(255), nullable=True)
    nombre_estudiante = Column(String(255), nullable=False)
    area_consulta = Column(String(100), nullable=True)
    
    # II. INFORMACIÓN GENERAL DEL CONSULTANTE
    remitido_por = Column(String(255), nullable=True)
    correo_electronico = Column(String(255), nullable=True)
    nombre_consultante = Column(String(255), nullable=False)
    edad = Column(Integer, nullable=True)
    fecha_nacimiento_dia = Column(Integer, nullable=True)
    fecha_nacimiento_mes = Column(Integer, nullable=True)
    fecha_nacimiento_ano = Column(Integer, nullable=True)
    lugar_nacimiento = Column(String(255), nullable=True)
    sexo = Column(String(20), nullable=True)  # Femenino/Masculino
    tipo_documento = Column(String(10), nullable=True)  # T.I./C.C./NUIP
    numero_documento = Column(String(20), nullable=False)
    lugar_expedicion = Column(String(255), nullable=True)
    direccion = Column(String(255), nullable=True)
    barrio = Column(String(100), nullable=True)
    estrato = Column(Integer, nullable=True)
    numero_telefonico = Column(String(20), nullable=True)
    numero_celular = Column(String(20), nullable=True)
    estado_civil = Column(String(50), nullable=True)
    escolaridad = Column(String(100), nullable=True)
    profesion_oficio = Column(String(100), nullable=True)
    
    # III. BREVE DESCRIPCIÓN DEL CASO
    descripcion_caso = Column(Text, nullable=True)
    
    # IV. CONCEPTO DEL ESTUDIANTE
    concepto_estudiante = Column(Text, nullable=True)
    
    # V. CONCEPTO DEL ASESOR JURÍDICO
    concepto_asesor = Column(Text, nullable=True)
    
    # VI. DECLARACIÓN DEL USUARIO (texto fijo)
    # Este campo se llenará automáticamente con el texto estándar
    
    # Campos de control
    activo = Column(Boolean, nullable=False, default=True)
    created_at = Column(DateTime, nullable=False, default=get_colombia_time)
    updated_at = Column(DateTime, nullable=False, default=get_colombia_time, onupdate=get_colombia_time)
    created_by = Column(Integer, ForeignKey("users.id"), nullable=False)

    # Relaciones
    user = relationship("User", back_populates="controles_operativos")
    documentos = relationship("Documento", back_populates="control_operativo")

    def __repr__(self):
        return f"<ControlOperativo(id={self.id}, estudiante='{self.nombre_estudiante}', fecha={self.fecha_dia}/{self.fecha_mes}/{self.fecha_ano})>"

    def to_dict(self):
        """Convierte el objeto a diccionario para API"""
        return {
            "id": self.id,
            # I. DATOS DEL USUARIO
            "ciudad": self.ciudad,
            "fecha_dia": self.fecha_dia,
            "fecha_mes": self.fecha_mes,
            "fecha_ano": self.fecha_ano,
            "nombre_docente_responsable": self.nombre_docente_responsable,
            "nombre_estudiante": self.nombre_estudiante,
            "area_consulta": self.area_consulta,
            # II. INFORMACIÓN GENERAL DEL CONSULTANTE
            "remitido_por": self.remitido_por,
            "correo_electronico": self.correo_electronico,
            "nombre_consultante": self.nombre_consultante,
            "edad": self.edad,
            "fecha_nacimiento_dia": self.fecha_nacimiento_dia,
            "fecha_nacimiento_mes": self.fecha_nacimiento_mes,
            "fecha_nacimiento_ano": self.fecha_nacimiento_ano,
            "lugar_nacimiento": self.lugar_nacimiento,
            "sexo": self.sexo,
            "tipo_documento": self.tipo_documento,
            "numero_documento": self.numero_documento,
            "lugar_expedicion": self.lugar_expedicion,
            "direccion": self.direccion,
            "barrio": self.barrio,
            "estrato": self.estrato,
            "numero_telefonico": self.numero_telefonico,
            "numero_celular": self.numero_celular,
            "estado_civil": self.estado_civil,
            "escolaridad": self.escolaridad,
            "profesion_oficio": self.profesion_oficio,
            # III. BREVE DESCRIPCIÓN DEL CASO
            "descripcion_caso": self.descripcion_caso,
            # IV. CONCEPTO DEL ESTUDIANTE
            "concepto_estudiante": self.concepto_estudiante,
            # V. CONCEPTO DEL ASESOR JURÍDICO
            "concepto_asesor": self.concepto_asesor,
            # Campos de control
            "activo": self.activo,
            "created_at": self.created_at.isoformat() if self.created_at else None,
            "updated_at": self.updated_at.isoformat() if self.updated_at else None,
            "created_by": self.created_by
        }

File: go-backend/internal/models/control_operativo.go
package models

import (
	"time"
)

type ControlOperativo struct {
	ID                       uint      `gorm:"primaryKey" json:"id"`
	Ciudad                   string    `gorm:"type:varchar(100)" json:"ciudad"`
	FechaDia                 int       `gorm:"not null" json:"fecha_dia"`
	FechaMes                 int       `gorm:"not null" json:"fecha_mes"`
	FechaAno                 int       `gorm:"not null" json:"fecha_ano"`
	NombreDocenteResponsable string    `gorm:"type:varchar(255)" json:"nombre_docente_responsable"`
	NombreEstudiante         string    `gorm:"type:varchar(255)" json:"nombre_estudiante"`
	AreaConsulta             string    `gorm:"type:varchar(100)" json:"area_consulta"`
	RemitidoPor              string    `gorm:"type:varchar(255)" json:"remitido_por"`
	CorreoElectronico        string    `gorm:"type:varchar(255)" json:"correo_electronico"`
	NombreConsultante        string    `gorm:"type:varchar(255)" json:"nombre_consultante"`
	Edad                     int       `json:"edad"`
	FechaNacimientoDia       int       `json:"fecha_nacimiento_dia"`
	FechaNacimientoMes       int       `json:"fecha_nacimiento_mes"`
	FechaNacimientoAno       int       `json:"fecha_nacimiento_ano"`
	LugarNacimiento          string    `gorm:"type:varchar(255)" json:"lugar_nacimiento"`
	Sexo                     string    `gorm:"type:varchar(20)" json:"sexo"`
	TipoDocumento            string    `gorm:"type:varchar(10)" json:"tipo_documento"`
	NumeroDocumento          string    `gorm:"type:varchar(20)" json:"numero_documento"`
	LugarExpedicion          string    `gorm:"type:varchar(255)" json:"lugar_expedicion"`
	Direccion                string    `gorm:"type:varchar(255)" json:"direccion"`
	Barrio                   string    `gorm:"type:varchar(100)" json:"barrio"`
	Estrato                  int       `json:"estrato"`
	NumeroTelefonico         string    `gorm:"type:varchar(20)" json:"numero_telefonico"`
	NumeroCelular            string    `gorm:"type:varchar(20)" json:"numero_celular"`
	EstadoCivil              string    `gorm:"type:varchar(50)" json:"estado_civil"`
	Escolaridad              string    `gorm:"type:varchar(100)" json:"escolaridad"`
	ProfesionOficio          string    `gorm:"type:varchar(100)" json:"profesion_oficio"`
	DescripcionCaso          string    `gorm:"type:text" json:"descripcion_caso"`
	ConceptoEstudiante       string    `gorm:"type:text" json:"concepto_estudiante"`
	ConceptoAsesor           string    `gorm:"type:text" json:"concepto_asesor"`
	EstadoFlujo              string    `gorm:"type:varchar(50);default:'pendiente_profesor'" json:"estado_flujo"`
	EstadoResultado          *string   `gorm:"type:varchar(50)" json:"estado_resultado"`
	Activo                   bool      `gorm:"default:true" json:"activo"`
	CreatedAt                time.Time `json:"created_at"`
	UpdatedAt                time.Time `json:"updated_at"`
	CreatedByID              uint      `gorm:"not null" json:"created_by"`
	CreatedBy                User      `gorm:"foreignKey:CreatedByID" json:"created_by_user,omitempty"`
	DocumentosAdjuntos       []DocumentoAdjunto `gorm:"foreignKey:ControlOperativoID" json:"documentos_adjuntos,omitempty"`
	Notificaciones           []Notificacion     `gorm:"foreignKey:ControlOperativoID" json:"notificaciones,omitempty"`
}

type DocumentoAdjunto struct {
	ID                 uint      `gorm:"primaryKey" json:"id"`
	ControlOperativoID uint      `gorm:"not null" json:"control_operativo_id"`
	NombreOriginal     string    `gorm:"type:varchar(255)" json:"nombre_original"`
	NombreArchivo      string    `gorm:"type:varchar(255)" json:"nombre_archivo"`
	TipoArchivo        string    `gorm:"type:varchar(50)" json:"tipo_archivo"`
	TamanoBytes        int64     `json:"tamano_bytes"`
	RutaArchivo        string    `gorm:"type:varchar(500)" json:"ruta_archivo"`
	ConvertidoPDF      bool      `gorm:"default:false" json:"convertido_pdf"`
	RutaPDF            string    `gorm:"type:varchar(500)" json:"ruta_pdf"`
	CreatedAt          time.Time `json:"created_at"`
	UpdatedAt          time.Time `json:"updated_at"`
}

type Notificacion struct {
	ID                 uint      `gorm:"primaryKey" json:"id"`
	ControlOperativoID uint      `gorm:"not null" json:"control_operativo_id"`
	UserID             uint      `gorm:"not null" json:"user_id"`
	TipoNotificacion   string    `gorm:"type:varchar(100)" json:"tipo_notificacion"`
	Mensaje            string    `gorm:"type:text" json:"mensaje"`
	Leida              bool      `gorm:"default:false" json:"leida"`
	CreatedAt          time.Time `json:"created_at"`
	User               User      `gorm:"foreignKey:UserID" json:"user,omitempty"`
	ControlOperativo   ControlOperativo `gorm:"foreignKey:ControlOperativoID" json:"control_operativo,omitempty"`
}

// DTOs para requests
type ControlOperativoRequest struct {
	Ciudad                   string `json:"ciudad" binding:"required"`
	FechaDia                 int    `json:"fecha_dia" binding:"required"`
	FechaMes                 int    `json:"fecha_mes" binding:"required"`
	FechaAno                 int    `json:"fecha_ano" binding:"required"`
	NombreDocenteResponsable string `json:"nombre_docente_responsable" binding:"required"`
	NombreEstudiante         string `json:"nombre_estudiante" binding:"required"`
	AreaConsulta             string `json:"area_consulta" binding:"required"`
	RemitidoPor              string `json:"remitido_por"`
	CorreoElectronico        string `json:"correo_electronico"`
	NombreConsultante        string `json:"nombre_consultante" binding:"required"`
	Edad                     int    `json:"edad"`
	FechaNacimientoDia       int    `json:"fecha_nacimiento_dia"`
	FechaNacimientoMes       int    `json:"fecha_nacimiento_mes"`
	FechaNacimientoAno       int    `json:"fecha_nacimiento_ano"`
	LugarNacimiento          string `json:"lugar_nacimiento"`
	Sexo                     string `json:"sexo"`
	TipoDocumento            string `json:"tipo_documento"`
	NumeroDocumento          string `json:"numero_documento"`
	LugarExpedicion          string `json:"lugar_expedicion"`
	Direccion                string `json:"direccion"`
	Barrio                   string `json:"barrio"`
	Estrato                  int    `json:"estrato"`
	NumeroTelefonico         string `json:"numero_telefonico"`
	NumeroCelular            string `json:"numero_celular"`
	EstadoCivil              string `json:"estado_civil"`
	Escolaridad              string `json:"escolaridad"`
	ProfesionOficio          string `json:"profesion_oficio"`
	DescripcionCaso          string `json:"descripcion_caso" binding:"required"`
	ConceptoEstudiante       string `json:"concepto_estudiante" binding:"required"`
}

type ConceptoAsesorRequest struct {
	ConceptoAsesor string `json:"concepto_asesor" binding:"required"`
}

type EstadoResultadoRequest struct {
	EstadoResultado string `json:"estado_resultado" binding:"required"`
}